contracts/starknet/space/space.cairo

```diff
@@ -0,0 +1,134 @@
+%lang starknet
+
+from starkware.starknet.common.syscalls import ( get_caller_address, get_block_number )
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.cairo.common.math import assert_lt, assert_le, assert_nn, assert_not_zero
+
+struct Vote:
+    member choice: felt # TODO use Choice enum
+    member eth_address: felt # TODO: use address
+    member voting_power: felt
+end
+
+struct Proposal:
+    member execution_hash: felt # TODO: Use Hash type
+    member start_block: felt
+    member end_block: felt
+end
+
+# TODO: use L1Address instead of felt
+@contract_interface
+namespace IVotingStrategy:
+    func get_voting_power(address: felt, at: felt) -> (voting_power: felt):
+    end
+end
+
+@storage_var
+func voting_delay() -> (delay : felt):
+end
+
+
+@storage_var
+func voting_period() -> (period : felt):
+end
+
+@storage_var
+func proposal_threshold() -> (threshold : felt):
+end
+
+# TODO: Should be Address not felt
+@storage_var
+func voting_strategy() -> (strategy_address : felt):
+end
+
+@storage_var
+func authenticator() -> (authenticator_address : felt):
+end
+
+@storage_var
+func next_proposal_nonce() -> (nonce : felt):
+end
+
+@storage_var
+func proposals(proposal_id: felt) -> (proposal : Proposal):
+end
+
+@storage_var
+func votes(proposal_id: felt) -> (vote : Vote):
+end
+
+@constructor
+func constructor{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr: felt}(_voting_delay: felt, _voting_period: felt, _proposal_threshold: felt, _voting_strategy: felt, _authenticator: felt):
+    # Sanity checks
+    assert_nn(_voting_delay)
+    assert_nn(_voting_period)
+    assert_nn(_proposal_threshold)
+    assert_not_zero(_voting_strategy)
+    assert_not_zero(_authenticator)
+
+    # Initialize the storage variables
+    voting_delay.write(_voting_delay)
+    voting_period.write(_voting_period)
+    proposal_threshold.write(_proposal_threshold)
+    voting_strategy.write(_voting_strategy)
+    authenticator.write(_authenticator)
+
+    return ()
+end
+
+# TODO: should be address not felt, and choice should be of enum Choice not felt
+@external
+func vote{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr: felt}(eth_address: felt, proposal_id: felt, choice: felt) -> ():
+
+    # Verify that the caller is the authenticator contract.
+    let (caller_address) = get_caller_address()
```

Comment:
Consider defining a function called authenticator_only that checks the caller is the authenticator. Something that looks like a modifier in solidity. We do something similar in [Starkgate](https://github.com/starkware-libs/starkgate-contracts/blob/02f61b33b5177d6d75c6ee20c61d22d3cd97de9f/src/starkware/starknet/apps/starkgate/cairo/ERC20.cairo#L107)